sr_robot_lib/src/sr_motor_robot_lib.cpp

```diff
@@ -1197,100 +1194,110 @@ namespace shadow_robot
+    this->nullify_demand_ = true;
+
+    control_type_ = control_type;
 
-    if (control_type == sr_robot_msgs::ControlType::PWM)
+    std::string controller_from_suffix, controller_to_suffix;
+    if (control_type_.control_type == sr_robot_msgs::ControlType::PWM)
     {
-      env_variable = "PWM_CONTROL=1";
-      param_value = "PWM";
+      ROS_INFO("Changing control type to PWM");
+      controller_from_suffix = "_effort_controller";
+      controller_to_suffix = "_position_controller";
     }
     else
     {
-      env_variable = "PWM_CONTROL=0";
-      param_value = "FORCE";
+      ROS_INFO("Changing control type to TORQUE");
+      controller_from_suffix = "_position_controller";
+      controller_to_suffix = "_effort_controller";
     }
-
-    string arguments = "";
-
-    // Read the hand_id prefix from the parameter server
-    // The hand_id will be passed as an argument to the sr_edc_default_controllers.launch
-    // so that the parameters in it will be read from the correct files
-    string hand_id = "";
-    this->nodehandle_.template param<string>("hand_id", hand_id, "");
-    ROS_DEBUG("hand_id: %s", hand_id.c_str());
-    arguments += " hand_id:=" + hand_id;
-    ROS_INFO("arguments: %s", arguments.c_str());
-
-    int result = system((env_variable + " roslaunch sr_ethercat_hand_config sr_edc_default_controllers.launch" +
-                         arguments).c_str());
-
-    if (result == 0)
+    std::thread client_thread([=]()
     {
-      ROS_WARN("New parameters loaded successfully on Parameter Server");
-
-      this->nh_tilde.setParam("default_control_mode", param_value);
-
-      // We need another node handle here, that is at the node's base namespace,
-      // as the controllers and the controller manager are unique per ethercat loop.
       ros::NodeHandle nh;
-      ros::ServiceClient list_ctrl_client = nh.template serviceClient<controller_manager_msgs::ListControllers>(
-              "controller_manager/list_controllers");
-      controller_manager_msgs::ListControllers controllers_list;
 
+      std::set<std::string> loaded_controllers;
+      ros::ServiceClient list_ctrl_client = nh.template serviceClient
+        <controller_manager_msgs::ListControllers>("controller_manager/list_controllers");
+      controller_manager_msgs::ListControllers controllers_list;
       if (list_ctrl_client.call(controllers_list))
       {
-        for (unsigned int i = 0; i < controllers_list.response.controller.size(); ++i)
+        for (auto controller : controllers_list.response.controller)
         {
-          if (ros::service::exists(controllers_list.response.controller[i].name + "/reset_gains", false))
-          {
-            ros::ServiceClient reset_gains_client = nh.template serviceClient<std_srvs::Empty>(
-              controllers_list.response.controller[i].name + "/reset_gains");
-            std_srvs::Empty empty_message;
-            if (!reset_gains_client.call(empty_message))
```

Comment:
Do we know what `reset_gains` did, and thus why we don't need to call it anymore?

Reply:
Resetting gains reads parameters (like `max_force` and `friction_deadband`). The same parameters are being read on controller starting method so resetting should be unnecessary when starting stopped controller.